README.md
Calamity is a publish-subscribe event bus library, allowing for the building of loosely-coupled, event-driven applications.
The event bus can either be used globally or it can be attached to individual objects, providing any object with the ability to publish local events.

Install via NPM:

    npm install --save calamity

Or via Bower:

    bower install --save calamity

# Local event busses with `emitter()`.
Local event busses are used on individual classes and objects to allow them to publish events others can subscribe to.
When attaching a local event bus, two methods are made available: `on()` and `trigger()`.

## `on(address, handler(msg) )`
Attaches the function `handler(msg)` to be executed when `address` is triggered.
When `handler` is executed, Calamity will automatically bind its execution to the subscribing object.

## `trigger(address [, data [, replier(reply) ] ] )`
Triggers any handlers attached to `address`.
`data` is an arbitrary object which will be provided to the handler via its message.
`replier` is a secondary handler function which will be executed in case the main handler executes a reply (more on this later).

## Local events example
Consider this imaginary `View` class:

    :::coffeescript
    Calamity = require "calamity"
    _ = require "underscore"

    class View
    	Calamity.emitter @prototype

    	render:
    		@trigger "beforeRender"
    		_.defer =>
    			# Real render code ...
    			@element.find("#closeButton").click =>
    				@trigger "closeButtonClicked"
    		@trigger "afterRender"

Now using the `on()` method, any code with a reference to the view can subscribe to events coming out of it.
In this example we can act on the view starting and finishing its render routine.
Additionally, the view attaches to the click event on a DOM element, propagating it to a Calamity event on the view itself, thereby abstracting it away.

    :::coffeescript
    view = new View()
    view.on "closeButtonClicked", ->
    	view.hide()

# Global event bus with `proxy()`
A more powerful example of Calamity is using it as a global event bus which spans your entire application.
The global event bus proxy attaches three methods: `subscribe()`, `publish()` and `send()`.

## `subscribe( address, handler(msg) )`
Attaches the function `handler(msg)` to be executed when messages are published or sent to `address`.
When `handler` is executed, Calamity will automatically bind its execution to the subscribing object.

## `publish( address [, data [, replier(reply) ] ] )`
Sends a message to `address`.
`data` is an arbitrary object which will be provided to the handler via its message.
`replier` is a secondary handler function which will be executed in case the main handler executes a reply (more on this later).

## `send( address [, data [, replier(reply) ] ] )`
Works exactly like `publish()` with the notable difference that only one subscribing handler will be executed.
This is useful for sending commands instead events.

## Global events example

    :::coffeescript
    Calamity = require "calamity"
    class Foo
    	Calamity.proxy @prototype

    	constructor: () ->
    		@subscribe "address", @handler

    	handler: (msg) ->
    		# Proxy automatically handles binding to this.
    		@data = msg.data

Now, whenever any object publishes a message to the `foo:bar` address, handler will be called and you can react on it.

# Events and commands
*Events* tell of something which happened and are transmitted using `trigger()` and `publish()`.
Normally events will not be replied to and will originate from a single source.

*Commands* are instructions to perform an action and are transmitted using `send()`.
The local events mixin does not include a mechanism for sending commands.
Commands are much more likely to send replies than events, as they often need to report on their progress.
Commands will usually only be subscribed to by a single object type, and be sent from many.

Within Calamity there is no real difference between an event and a command, other than the method used for sending it.
Both of them use the `EventMessage` class to contain themselves.

# Working with messages
When subscribing to an address, the attached handler is expected to accept a single argument: an `EventMessage` object.
This object encapsulates the entire message and everything that can be done with it.

## EventMessage properties
The following properties are available on `EventMessage`:

* `msg.address`: The address the message was sent to.
* `msg.data`: The raw data object supplied on creation.
* `msg.status`: The status of the message. This can beeither "ok" or "error". Error is used by `replyError()`. To check for erros, `isError()` should be used.
* `msg.error`: If `msg.status` is "error", this will contain the error supplied to `replyError()`.

## EventMessage methods
The following methods can be used to interact with the message:

* `getOptional( param, default )`: Returns a value from `msg.data`. If the value does not exist, the supplied `default` is returned.
* `getRequried( param )`: Returns a value from `msg.data`. IF the value does not exist, an `Error` is thrown.
* `reply( data[, replier(msg) ] )`: Sends a reply to the sender.
* `replyError( error[, replier(msg) ] )`: Sends an error reply to the sender.
* `isError()`: Returns true if the message represents an error. The supplied error is available via `msg.error`.

# Replying to messages

# License
Calamity is licensed and freely distributed under the [MIT License][mit]

[mit]: https://bitbucket.org/kennethjor/calamity/raw/default/LICENSE "MIT License"


CHANGELOG.md
# Changelog

## 0.5.0 (dev)
* *API change:* Replies now send a new `EventMessage` object instead of plain object.
* *API change:* Changed the way `EventMessage.replyError` serializes the error object to something more meaningful.
* *API change:* Changed events to fire synchronously as an experiment. Asynchronous handling seems to cause more trouble that it's worth.
* *API change:* Removed `calamity.proxy()`, it didn't really make sense.
* *API change:* Renamed `EventBus` to `Bus`.
* *API change:* Renamed `EventMessage` to `Message`.
* *API change:* Rewrote `Bridge` to serve as a basic bridge that can be extended.
* *Feature:* Implemented `replyError()` on `EventMessage` to allow for easy handling of errors.
* *Feature:* Implemented `toJSON()`  and `fromJSON()` on `EventMessage`.
* *Feature:* Implemented `getRequired()`  and `getOptional()` on `EventMessage` for data retrieval.
* *Feature:* Implemented `catch()` on `EventMessage` to handle automatic catching of asynchronous errors. ([pull/2][pull/2])
* *Feature:* `getOptional` and `getRequired` on `EventMessage` can now return deep values using JSON dot syntax.
* *Fix:* `EventMessages` with empty datasets would throw undefined errors with `getRequired` and `getOptional`.

## 0.4.1 (2013-07-10)
* *Fix:* `Subscription`'s `unsubscribe()` contained a typo, causing it to never work.

## 0.4.0 (2013-06-26)
* *Feature:* Added support for commands via `send()`.
* *Feature:* Added support for basic event bridges via `bridge()`.

## 0.3.1 (2013-03-19)
* *Fix:* Added AMD load header. ([pull/1][pull/1])

## 0.3.0 (2013-01-29)
* *API change:* Removed initial underscore from `_trigger()`, `_subscribe()`, and `_publish()` on `ProxyMixin` and `EmitterMixin`. It makes assumptions about code style.
* *Fix:* Browser initialisation.
* *Fix:* Bug where `EmitterMixin` would not correctly set default context.

## 0.2.0 (2012-11-14)

* *Feature:* Implemented `ProxyMixin` for easy global event bus handling.
* *Feature:* Implemented `Subscription` objects.
* *Feature:* Implemented `EmitterMixin` for easy attachment of an instance-local `EventBus`.

## 0.1.1 (2012-11-12)

* `npm` related fixes.

## 0.1.0 (2012-11-12)

* Initial public release.
* *Feature:* `EventBus` implementation.

[pull/1]: https://bitbucket.org/kennethjor/calamity/pull-request/1 "Added RequireJS support to init.js."
[pull/2]: https://bitbucket.org/kennethjor/calamity/pull-request/2 "Error proxy."


spec/core/EventBusTest.js
// Generated by CoffeeScript 1.4.0
(function() {
  var Calamity;

  Calamity = require("../../../calamity.js");

  describe("EventBus", function() {
    it("should create an empty message", function() {
      var bus, msg;
      bus = new EventBus();
      msg = bus._createMessage();
      return expect(msg instanceof EventMessage).toBe(true);
    });
    it("should create a message with correct address, data, and reply", function() {
      var bus, msg, reply;
      bus = new EventBus();
      reply = function() {};
      msg = bus._createMessage("address", "data", reply);
      expect(msg instanceof EventMessage).toBe(true);
      expect(msg.address).toBe("address");
      expect(msg.data).toBe("data");
      return expect(msg._replyHandler).toBe(reply);
    });
    return it("should create multiple messages", function() {
      var bus, msg1, msg2;
      bus = new EventBus();
      msg1 = bus._createMessage();
      msg2 = bus._createMessage();
      return expect(msg1).toNotBe(msg2);
    });
  });

}).call(this);


grunt.js
module.exports = function(grunt) {

	// Project configuration.
	grunt.initConfig({
		pkg: '<json:package.json>',
		meta: {
			banner: "/*! <%= pkg.fullname %> <%= pkg.version %> - MIT license */",
			wrapperVersion: "C.VERSION = \"<%= pkg.version %>\";",
			wrapperStart: "(function(){",
			wrapperEnd: "}).call(this);"
		},

		coffee: {
			// Initial precompile to catch compile errors properly.
			core_first: {
				files: {
					"build/core/*.js": ["src/core/*.coffee"]
				},
				options: {
					bare: true
				}
			},
			// Second compile of concatenated coffee files for more minimizes output.
			core_second: {
				files: {
					"build/core/calamity.js": ["build/core/<%= pkg.name %>.coffee"]
				},
				options: {
					bare: true
				}
			},
			// Compile the core test coffee files.
			core_test: {
				files: {
					"build/test/core/*.js": ["test/core/*.coffee"]
				}
			}
		},

		// Build Jasmine core specs for the browser
		browserify: {
			"build/spec.js": {
				entries: ["spec/core/*.coffee"]
			}
		},

		// Execute Jasmine specs
		jasmine : {
			src: [
				"node_modules/sinon/lib/sinon.js",
				"node_modules/sinon/lib/sinon/spy.js",
				"lib/jasmine-sinon.js"
			],
			specs : "build/spec.js",
			timeout : 500,
			verbose: true,
			junit : {
				output : "build/spec/"
			},
			phantomjs : {
				"ignore-ssl-errors" : true
			}
		},

		concat: {
			// Assembles the core coffee files in prep for full compile.
			core_coffee: {
				src: "src/core/*.coffee",
				dest: "build/core/<%= pkg.name %>.coffee"
			},
			// Assembles the core distribution files.
			core_dist: {
				src: [
					"<banner>",
					"<banner:meta.wrapperStart>",
					"src/init/init.js",
					"<banner:meta.wrapperVersion>",
					"<%= _.keys(coffee.core_second.files)[0] %>",
					"<banner:meta.wrapperEnd>"
				],
				dest: "<%= pkg.name %>.js"
			}
		},
		min: {
			// Minimizes the core distribution files.
			core: {
				src: [
					"<banner>",
					"<config:concat.core_dist.dest>"
				],
				dest: "<%= pkg.name %>-min.js"
			}
		},

		test: {
			files : [
				// Core tests.
				"<%= _.keys(coffee.core_test.files) %>"
			]
		},

		watch: {
			files: [
				"src/**",
				"test/**"
			],
			tasks: "default"
		}
	});

	// Load grunt plugins.
	grunt.loadNpmTasks("grunt-contrib-coffee");
	grunt.loadNpmTasks("grunt-jasmine-runner");
//	grunt.loadNpmTasks("grunt-jasmine-node");
	grunt.loadNpmTasks("grunt-browserify");

	// Core compile.
	grunt.registerTask("compile-core", "coffee:core_first concat:core_coffee coffee:core_second");
	grunt.registerTask("dist-core", "concat:core_dist min:core");
	grunt.registerTask("test-core", "coffee:core_test test"); // will go away once ported to jasmine
	grunt.registerTask("spec-phantomjs", "browserify jasmine");
	grunt.registerTask("build-core", "compile-core dist-core test-core spec-phantomjs");
	// Default task.
	grunt.registerTask("default", "build-core");

};


Gruntfile.js
var exec = require("child_process").exec;

module.exports = function(grunt) {
	// Project configuration.
	grunt.initConfig({
		pkg: grunt.file.readJSON("package.json"),

		coffee: {
			// Core compile
			core: {
				files: {
					"build/calamity.js": ["src/core/*.coffee"]
				},
				options: {
					bare: true
				}
			},
			// Compile the test files.
			test: {
				expand: true,
				cwd: "test",
				src: ["**/*.coffee"],
				dest: "build/test",
				ext: ".js"
			},
			// Compile the Jasmine specs.
			spec: {
				expand: true,
				cwd: "spec",
				src: ["**/*.coffee"],
				dest: "build/spec",
				ext: ".js"
			}
		},

		concat: {
			// Assembles the core distribution files.
			core: {
				options: {
					banner:
						"/*! <%= pkg.fullname %> <%= pkg.version %> - MIT license */\n" +
						"(function(){\n",
					footer: "}).call(this);",
					process: true
				},
				src: [
					"src/init/init.js",
					"<%= _.keys(coffee.core.files)[0] %>"
				],
				dest: "<%= pkg.name %>.js"
			}
		},

		uglify: {
			core: {
				options: {
					preserveComments: "some"
				},
				files: {
					"calamity-min.js": "calamity.js"
				}
			}
		},

		nodeunit: {
			all: ["build/test/**/*Test.js"]
		},

		watch: {
			files: [
				"src/**",
				"test/**",
				"spec/**"
			],
			tasks: "default"
		}
	});

	// Jessie Jasmine runner.
	grunt.registerTask("jessie", "Runs Jasmine with Jessie.", function() {
		done = this.async();
		command = "./node_modules/jessie/bin/jessie build/spec"
		exec(command, function(err, stdout, stderr) {
			console.log(stdout);
			if (err) {
				grunt.warn(err);
				done(false);
			}
			else {
				done(true);
			}
		});
	});

	// Load grunt plugins.
	grunt.loadNpmTasks("grunt-contrib-coffee");
	grunt.loadNpmTasks("grunt-contrib-concat");
	grunt.loadNpmTasks("grunt-contrib-uglify");
	grunt.loadNpmTasks("grunt-contrib-nodeunit");

	// Default task.
	grunt.registerTask("default", ["coffee", "concat", "uglify", "nodeunit", "jessie"]);
};


calamity-min.js
/*! Calamity 0.5.0-rc.7 - MIT license */
(function(){"undefined"==typeof _&&"function"==typeof require&&(_=require("underscore"));var t={version:"0.5.0-rc.7"},i=this;"undefined"!=typeof exports?("undefined"!=typeof module&&module.exports&&(exports=module.exports=t),exports.Calamity=t):"function"==typeof define&&define.amd?define(["calamity"],t):i.calamity=t;var r,s,e,n,o,u,h,a,c,p,l,d,f=[].slice,b={}.hasOwnProperty;r=t.Bridge=function(){function t(){var t,i,r,s;for(i=1<=arguments.length?f.call(arguments,0):[],this._seen={},this._busses=i,r=0,s=i.length;s>r;r++)t=i[r],this.subscribeBus(t)}return t.prototype.SEEN_TIME=500,t.prototype._busses=null,t.prototype._seen=null,t.prototype._cleanId=null,t.prototype.subscribeBus=function(t){t.subscribe("*",function(t){return function(i){return function(r){return t.handle(i,r)}}}(this)(t))},t.prototype.handle=function(t,i){var r,s,e,n;if(!this.seen(i))for(n=this._busses,s=0,e=n.length;e>s;s++)r=n[s],r!==t&&r.publish(i)},t.prototype.seen=function(t,i){var r,s;return null==i&&(i=!0),r=(new Date).getTime()-this.SEEN_TIME,s=this._seen[t.id],null!=s&&s>r?!0:(i&&(this._seen[t.id]=(new Date).getTime(),this._scheduleClean()),!1)},t.prototype._scheduleClean=function(){this._cleanId||(this._cleanId=_.delay(function(t){return function(){t._clean(),_.isEmpty(t._seen)||t._scheduleClean()}}(this),this.SEEN_TIME))},t.prototype._clean=function(){var t,i,r,s;r=this._seen,i=(new Date).getTime()-this.SEEN_TIME;for(t in r)b.call(r,t)&&(s=r[t],i>s&&delete r[t])},t}(),s=t.Bus=function(){function t(){this.id=d.genId(),this._subscriptions={},this._bridges=[]}return t.prototype.subscribe=function(t,i,r){var s;return this._subscriptions[t]||(this._subscriptions[t]=[]),s=new h(t,i,r,this),this._subscriptions[t].push(s),this._bridgeProp("subscribe",{subscription:s}),s},t.prototype.unsubscribe=function(t,i){var r,s,e,n,o,u,a,c,p;if(e=t,e instanceof h){if(t=e.address,!this._subscriptions[t])return;for(c=this._subscriptions[t],r=n=0,u=c.length;u>n;r=++n)s=c[r],s===e&&this._subscriptions[t].splice(r)}else{if(!this._subscriptions[t])return;for(p=this._subscriptions[t],r=o=0,a=p.length;a>o;r=++o)s=p[r],s.address===t&&s.handler===i&&(e=s,this._subscriptions[t].splice(r))}this._bridgeProp("unsubscribe",{subscription:e})},t.prototype.publish=function(t,i,r){var s;return s=this._createMessage(t,i,r),t=s.address,s.sawBus(this)?this:(s.addBus(this),this._publishAddress(t,s),this._publishAddress("*",s),this._bridgeProp("publish",{message:s}),this)},t.prototype.send=function(t,i,r){var s;return s=this._createMessage(t,i,r),t=s.address,s.sawBus(this)?this:(s.addBus(this),this._sendAddress(t,s),this._bridgeProp("send",{message:s}),this)},t.prototype.bridge=function(t){if(!(t instanceof EventBridge))throw new Error("Briges must extend Calamity.EventBridge");return _.contains(this._bridges,t)||this._bridges.push(t),this},t.prototype._createMessage=function(t,i,r){var s;return s=t,s instanceof u||(s=new u(t,i,r)),s},t.prototype._publishAddress=function(t,i){var r,s,e,n;if(this._subscriptions[t])for(n=this._subscriptions[t],s=0,e=n.length;e>s;s++)r=n[s],r.trigger(i)},t.prototype._sendAddress=function(t,i){var r,s,e;this._subscriptions[t]&&(e=this._subscriptions[t],s=e.length,r=Math.floor(Math.random()*s),e[r].trigger(i))},t.prototype._bridgeProp=function(t,i){var r,s,e,n,o;if(this._bridges.length>0)for(r="bus."+t,i.bus=this,o=this._bridges,e=0,n=o.length;n>e;e++)s=o[e],s.trigger(r,i)},t}(),n=null,t.global=function(){return n||(n=new s),n},e=t.Emitter=function(){function t(){}return t.prototype.on=function(t,i,r){return r||(r=this),c(this).subscribe(t,i,r)},t.prototype.off=function(t,i,r){return p(this)?(r||(r=this),c(this).unsubscribe(t,i,r)):void 0},t.prototype.trigger=function(t,i,r){return p(this)?c(this).publish(t,i,r):void 0},t}(),p=function(t){var i,r;return null==(null!=t&&null!=(i=t._calamity)&&null!=(r=i.emitter)?r.bus:void 0)?!1:!0},c=function(t){var i,r;return i=t._calamity||(t._calamity={}),r=i.emitter||(i.emitter={}),r.bus||(r.bus=new s)},t.emitter=function(t){return _.extend(t,e.prototype)},u=t.Message=function(){function i(t,i,r){if(this.address=t,this.data=null!=i?i:{},this.id=d.genId(),this._busses=[],!_.isUndefined(r)&&!_.isFunction(r))throw new Error("Reply must be a function");this._replyHandler=r,this.status="ok",this.error=null}return i.prototype.reply=function(t,r){var s;return s=this._replyHandler,_.isFunction(s)?(t instanceof i||(t=new i(null,t,r)),s(t),this):void 0},i.prototype.replyError=function(t,r){var s,e,n,o,u,h;if(null==r&&(r={}),t instanceof Error){for(h="message,name,stack,fileName,lineNumber,description,number".split(","),o=0,u=h.length;u>o;o++)e=h[o],n=t[e],n&&"function"==typeof n.toString&&(n=n.toString()),r[e]=n;"function"==typeof t.toString&&(r.string=t.toString(),t=r.string,r.stack&&(t+=" :: "+r.stack))}return s=new i(null,r),s.status="error",s.error=t,this.reply(s),this},i.prototype["catch"]=function(t,r){var s;if(null==r){if(!_.isFunction(t))throw new Error("Supplied handler is not a function, "+typeof t+" supplied");r=t,t=void 0}if(_.isFunction(this._replyHandler)){if(null!=t){if(!(t instanceof i))return t instanceof Error||(t=new Error(t)),void this.replyError(t);if(t.isError())return void this.reply(t)}try{r(t)}catch(e){s=e,this.replyError(s)}}else{if(null!=t){if(!(t instanceof i))throw t instanceof Error||(t=new Error(t)),t;if(t.isError())throw t.error}r(t)}},i.prototype.isSuccess=function(){return"ok"===this.status},i.prototype.isError=function(){return"error"===this.status},i.prototype.getOptional=function(t,i){var r,s,e,n,o,u;if(s=t.split("."),e=this.data[s[0]],s.length>1)for(u=s.splice(1),n=0,o=u.length;o>n;n++){if(r=u[n],!_.isObject(e)||null==e[r]){e=void 0;break}e=e[r]}return"undefined"==typeof e?i:e},i.prototype.getRequired=function(t){var i;if(i=this.getOptional(t),"undefined"==typeof i)throw new Error('Variable "'+t+'" not found on message with address "'+this.address+'"');return i},i.prototype.addBus=function(t){return this.sawBus(t)?this:(this._busses.push(t.id),this)},i.prototype.sawBus=function(t){return _.contains(this._busses,t.id)},i.prototype.toJSON=function(){var i;return i={calamity:t.version,address:this.address,data:this.data,status:this.status,error:this.error},null!=this._replyHandler&&(i.reply=_.bind(this.reply,this)),i},i.fromJSON=function(t){var r;if(!_.isObject(t))throw new Error("JSON must be an object");if(null==t.calamity)throw new Error("Serialized JSON is not for calamity: "+JSON.stringify(t));return r=new i(t.address,t.data,t.reply),r.status=t.status,r.error=t.error,r},i}(),h=t.Subscription=function(){function t(t,i,r,s){this.address=t,this.handler=i,this.context=r,this.bus=s,this.id=d.genId(),this.active=!0}return t.prototype.unsubscribe=function(){return this.active?(this.bus.unsubscribe(this),this.active=!1,this):void 0},t.prototype.trigger=function(t){var i;return this.active?(i=_.bind(this.handler,this.context),i(t),this):this},t}(),l=Math.random,a=Math.floor,o="0123456789abcdef".split(""),d=t.util={genId:function(){var t,i,r;for(i="",t=r=1;32>=r;t=++r)i+=o[a(l()*o.length)];return i}}}).call(this);

calamity.js
/*! Calamity 0.5.0-rc.7 - MIT license */
(function(){
// Import underscore if necessary.
if (typeof _ === "undefined" && typeof require === "function") {
	_ = require("underscore");
}

// Init Calamity object.
var Calamity = {version: "0.5.0-rc.7"};

var root = this
// CommonJS
if (typeof exports !== "undefined") {
	if (typeof module !== "undefined" && module.exports) {
		exports = module.exports = Calamity;
	}
	exports.Calamity = Calamity;
}
// AMD
else if (typeof define === "function" && define.amd) {
    define(['calamity'], Calamity);
}
// Browser
else {
	root['calamity'] = Calamity;
}
var Bridge, Bus, Emitter, GLOBAL_BUS, HEX, Message, Subscription, floor, getEmitterBus, hasEmitterBus, random, util,
  __slice = [].slice,
  __hasProp = {}.hasOwnProperty;

Bridge = Calamity.Bridge = (function() {
  Bridge.prototype.SEEN_TIME = 500;

  Bridge.prototype._busses = null;

  Bridge.prototype._seen = null;

  Bridge.prototype._cleanId = null;

  function Bridge() {
    var bus, busses, _i, _len;
    busses = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
    this._seen = {};
    this._busses = busses;
    for (_i = 0, _len = busses.length; _i < _len; _i++) {
      bus = busses[_i];
      this.subscribeBus(bus);
    }
    return;
  }

  Bridge.prototype.subscribeBus = function(bus) {
    bus.subscribe("*", (function(_this) {
      return function(bus) {
        return function(msg) {
          return _this.handle(bus, msg);
        };
      };
    })(this)(bus));
  };

  Bridge.prototype.handle = function(bus, msg) {
    var b, _i, _len, _ref;
    if (this.seen(msg)) {
      return;
    }
    _ref = this._busses;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      b = _ref[_i];
      if (b !== bus) {
        b.publish(msg);
      }
    }
  };

  Bridge.prototype.seen = function(msg, save) {
    var limit, time;
    if (save == null) {
      save = true;
    }
    limit = new Date().getTime() - this.SEEN_TIME;
    time = this._seen[msg.id];
    if ((time != null) && time > limit) {
      return true;
    }
    if (save) {
      this._seen[msg.id] = new Date().getTime();
      this._scheduleClean();
    }
    return false;
  };

  Bridge.prototype._scheduleClean = function() {
    if (this._cleanId) {
      return;
    }
    this._cleanId = _.delay(((function(_this) {
      return function() {
        _this._clean();
        if (!_.isEmpty(_this._seen)) {
          _this._scheduleClean();
        }
      };
    })(this)), this.SEEN_TIME);
  };

  Bridge.prototype._clean = function() {
    var id, limit, seen, time;
    seen = this._seen;
    limit = new Date().getTime() - this.SEEN_TIME;
    for (id in seen) {
      if (!__hasProp.call(seen, id)) continue;
      time = seen[id];
      if (time < limit) {
        delete seen[id];
      }
    }
  };

  return Bridge;

})();

Bus = Calamity.Bus = (function() {
  function Bus() {
    this.id = util.genId();
    this._subscriptions = {};
    this._bridges = [];
  }

  Bus.prototype.subscribe = function(address, handler, context) {
    var sub;
    if (!this._subscriptions[address]) {
      this._subscriptions[address] = [];
    }
    sub = new Subscription(address, handler, context, this);
    this._subscriptions[address].push(sub);
    this._bridgeProp("subscribe", {
      subscription: sub
    });
    return sub;
  };

  Bus.prototype.unsubscribe = function(address, handler) {
    var i, s, sub, _i, _j, _len, _len1, _ref, _ref1;
    sub = address;
    if (sub instanceof Subscription) {
      address = sub.address;
      if (!this._subscriptions[address]) {
        return;
      }
      _ref = this._subscriptions[address];
      for (i = _i = 0, _len = _ref.length; _i < _len; i = ++_i) {
        s = _ref[i];
        if (s === sub) {
          this._subscriptions[address].splice(i);
        }
      }
    } else {
      if (!this._subscriptions[address]) {
        return;
      }
      _ref1 = this._subscriptions[address];
      for (i = _j = 0, _len1 = _ref1.length; _j < _len1; i = ++_j) {
        s = _ref1[i];
        if (s.address === address && s.handler === handler) {
          sub = s;
          this._subscriptions[address].splice(i);
        }
      }
    }
    this._bridgeProp("unsubscribe", {
      subscription: sub
    });
  };

  Bus.prototype.publish = function(address, data, reply) {
    var msg;
    msg = this._createMessage(address, data, reply);
    address = msg.address;
    if (msg.sawBus(this)) {
      return this;
    }
    msg.addBus(this);
    this._publishAddress(address, msg);
    this._publishAddress("*", msg);
    this._bridgeProp("publish", {
      message: msg
    });
    return this;
  };

  Bus.prototype.send = function(address, data, reply) {
    var msg;
    msg = this._createMessage(address, data, reply);
    address = msg.address;
    if (msg.sawBus(this)) {
      return this;
    }
    msg.addBus(this);
    this._sendAddress(address, msg);
    this._bridgeProp("send", {
      message: msg
    });
    return this;
  };

  Bus.prototype.bridge = function(bridge) {
    if (!(bridge instanceof EventBridge)) {
      throw new Error("Briges must extend Calamity.EventBridge");
    }
    if (!_.contains(this._bridges, bridge)) {
      this._bridges.push(bridge);
    }
    return this;
  };

  Bus.prototype._createMessage = function(address, data, reply) {
    var msg;
    msg = address;
    if (!(msg instanceof Message)) {
      msg = new Message(address, data, reply);
    }
    return msg;
  };

  Bus.prototype._publishAddress = function(address, msg) {
    var subscription, _i, _len, _ref;
    if (!this._subscriptions[address]) {
      return;
    }
    _ref = this._subscriptions[address];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      subscription = _ref[_i];
      subscription.trigger(msg);
    }
  };

  Bus.prototype._sendAddress = function(address, msg) {
    var i, len, subs;
    if (!this._subscriptions[address]) {
      return;
    }
    subs = this._subscriptions[address];
    len = subs.length;
    i = Math.floor(Math.random() * len);
    subs[i].trigger(msg);
  };

  Bus.prototype._bridgeProp = function(type, data) {
    var address, b, _i, _len, _ref;
    if (!(this._bridges.length > 0)) {
      return;
    }
    address = "bus." + type;
    data.bus = this;
    _ref = this._bridges;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      b = _ref[_i];
      b.trigger(address, data);
    }
  };

  return Bus;

})();

GLOBAL_BUS = null;

Calamity.global = function() {
  GLOBAL_BUS || (GLOBAL_BUS = new Bus());
  return GLOBAL_BUS;
};

Emitter = Calamity.Emitter = (function() {
  function Emitter() {}

  Emitter.prototype.on = function(address, handler, context) {
    context || (context = this);
    return getEmitterBus(this).subscribe(address, handler, context);
  };

  Emitter.prototype.off = function(address, handler, context) {
    if (!hasEmitterBus(this)) {
      return;
    }
    context || (context = this);
    return getEmitterBus(this).unsubscribe(address, handler, context);
  };

  Emitter.prototype.trigger = function(address, data, reply) {
    if (!hasEmitterBus(this)) {
      return;
    }
    return getEmitterBus(this).publish(address, data, reply);
  };

  return Emitter;

})();

hasEmitterBus = function(obj) {
  var _ref, _ref1;
  if ((obj != null ? (_ref = obj._calamity) != null ? (_ref1 = _ref.emitter) != null ? _ref1.bus : void 0 : void 0 : void 0) == null) {
    return false;
  }
  return true;
};

getEmitterBus = function(obj) {
  var calamity, emitter;
  calamity = (obj._calamity || (obj._calamity = {}));
  emitter = (calamity.emitter || (calamity.emitter = {}));
  return emitter.bus || (emitter.bus = new Bus());
};

Calamity.emitter = function(obj) {
  return _.extend(obj, Emitter.prototype);
};

Message = Calamity.Message = (function() {
  function Message(address, data, replyHandler) {
    this.address = address;
    this.data = data != null ? data : {};
    this.id = util.genId();
    this._busses = [];
    if (!(_.isUndefined(replyHandler) || _.isFunction(replyHandler))) {
      throw new Error("Reply must be a function");
    }
    this._replyHandler = replyHandler;
    this.status = "ok";
    this.error = null;
  }

  Message.prototype.reply = function(data, replier) {
    var replyHandler;
    replyHandler = this._replyHandler;
    if (!_.isFunction(replyHandler)) {
      return;
    }
    if (!(data instanceof Message)) {
      data = new Message(null, data, replier);
    }
    replyHandler(data);
    return this;
  };

  Message.prototype.replyError = function(error, data) {
    var msg, v, val, _i, _len, _ref;
    if (data == null) {
      data = {};
    }
    if (error instanceof Error) {
      _ref = "message,name,stack,fileName,lineNumber,description,number".split(",");
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        v = _ref[_i];
        val = error[v];
        if (val && typeof val.toString === "function") {
          val = val.toString();
        }
        data[v] = val;
      }
      if (typeof error.toString === "function") {
        data.string = error.toString();
        error = data.string;
        if (data.stack) {
          error += " :: " + data.stack;
        }
      }
    }
    msg = new Message(null, data);
    msg.status = "error";
    msg.error = error;
    this.reply(msg);
    return this;
  };

  Message.prototype["catch"] = function(other, handler) {
    var err;
    if (handler == null) {
      if (!_.isFunction(other)) {
        throw new Error("Supplied handler is not a function, " + (typeof other) + " supplied");
      }
      handler = other;
      other = void 0;
    }
    if (!_.isFunction(this._replyHandler)) {
      if (other != null) {
        if (other instanceof Message) {
          if (other.isError()) {
            throw other.error;
          }
        } else {
          if (!(other instanceof Error)) {
            other = new Error(other);
          }
          throw other;
        }
      }
      handler(other);
    } else {
      if (other != null) {
        if (other instanceof Message) {
          if (other.isError()) {
            this.reply(other);
            return;
          }
        } else {
          if (!(other instanceof Error)) {
            other = new Error(other);
          }
          this.replyError(other);
          return;
        }
      }
      try {
        handler(other);
      } catch (_error) {
        err = _error;
        this.replyError(err);
      }
    }
  };

  Message.prototype.isSuccess = function() {
    return this.status === "ok";
  };

  Message.prototype.isError = function() {
    return this.status === "error";
  };

  Message.prototype.getOptional = function(param, def) {
    var part, parts, val, _i, _len, _ref;
    parts = param.split(".");
    val = this.data[parts[0]];
    if (parts.length > 1) {
      _ref = parts.splice(1);
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        part = _ref[_i];
        if (_.isObject(val) && (val[part] != null)) {
          val = val[part];
        } else {
          val = void 0;
          break;
        }
      }
    }
    if (typeof val === "undefined") {
      return def;
    }
    return val;
  };

  Message.prototype.getRequired = function(param) {
    var val;
    val = this.getOptional(param);
    if (typeof val === "undefined") {
      throw new Error("Variable \"" + param + "\" not found on message with address \"" + this.address + "\"");
    }
    return val;
  };

  Message.prototype.addBus = function(bus) {
    if (this.sawBus(bus)) {
      return this;
    }
    this._busses.push(bus.id);
    return this;
  };

  Message.prototype.sawBus = function(bus) {
    return _.contains(this._busses, bus.id);
  };

  Message.prototype.toJSON = function() {
    var json;
    json = {
      calamity: Calamity.version,
      address: this.address,
      data: this.data,
      status: this.status,
      error: this.error
    };
    if (this._replyHandler != null) {
      json.reply = _.bind(this.reply, this);
    }
    return json;
  };

  Message.fromJSON = function(json) {
    var msg;
    if (!_.isObject(json)) {
      throw new Error("JSON must be an object");
    }
    if (json.calamity == null) {
      throw new Error("Serialized JSON is not for calamity: " + (JSON.stringify(json)));
    }
    msg = new Message(json.address, json.data, json.reply);
    msg.status = json.status;
    msg.error = json.error;
    return msg;
  };

  return Message;

})();

Subscription = Calamity.Subscription = (function() {
  function Subscription(address, handler, context, bus) {
    this.address = address;
    this.handler = handler;
    this.context = context;
    this.bus = bus;
    this.id = util.genId();
    this.active = true;
    return;
  }

  Subscription.prototype.unsubscribe = function() {
    if (!this.active) {
      return;
    }
    this.bus.unsubscribe(this);
    this.active = false;
    return this;
  };

  Subscription.prototype.trigger = function(msg) {
    var bound;
    if (!this.active) {
      return this;
    }
    bound = _.bind(this.handler, this.context);
    bound(msg);
    return this;
  };

  return Subscription;

})();

random = Math.random;

floor = Math.floor;

HEX = "0123456789abcdef".split("");

util = Calamity.util = {
  genId: function() {
    var i, id, _i;
    id = "";
    for (i = _i = 1; _i <= 32; i = ++_i) {
      id += HEX[floor(random() * HEX.length)];
    }
    return id;
  }
};
}).call(this);
//# sourceMappingURL=calamity.js.map